### PR TITLE
debug options plus pipe pfm dumping

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
EDIT to last status:

- refactored pfm writing for support of some prefixes like img counter, in/out markers written
- introduce `--dump-pipe <module>` option writing in & output data of a module
-  prettyfied [memory] output
- introduces `-d common` option switching on the most important -d options
- make sure the tmp directory to write pfm files is existing

### CODE REVIEW

File: src/common/darktable.c [hunk 1]
```diff
@@ -129,9 +129,10 @@ static int usage(const char *argv0)
   printf("  --cachedir <user cache directory>\n");
   printf("  --conf <key>=<value>\n");
   printf("  --configdir <user config directory>\n");
-  printf("  -d {all,act_on,cache,camctl,camsupport,control,dev,imageio,\n");
+  printf("  -d {act_on,cache,camctl,camsupport,control,dev,imageio,\n");
   printf("      input,ioporder,lighttable,lua,masks,memory,nan,opencl,params,\n");
-  printf("      perf,print,pwstorage,signal,sql,tiling,undo,verbose,pipe}\n");
+  printf("      perf,print,pwstorage,signal,sql,tiling,undo,verbose,pipe,\n");
+  printf("      all,common}\n");
   printf("  --d-signal <signal> \n");
```
Comment (TurboGit) [inline src/common/darktable.c:136]: Maybe worth noting that common = opencl + dev + masks + params + imageio + pipe ?

File: src/common/darktable.h [hunk 2]
```diff
@@ -274,7 +274,8 @@ typedef enum dt_debug_thread_t
   DT_DEBUG_TILING         = 1 << 23,
   DT_DEBUG_VERBOSE        = 1 << 24,
   DT_DEBUG_PIPE           = 1 << 25,
-  DT_DEBUG_ALL            = 0xffffffff & ~DT_DEBUG_VERBOSE
+  DT_DEBUG_ALL            = 0xffffffff & ~DT_DEBUG_VERBOSE,
+  DT_DEBUG_COMMON         = DT_DEBUG_OPENCL | DT_DEBUG_DEV | DT_DEBUG_MASKS | DT_DEBUG_PARAMS | DT_DEBUG_IMAGEIO | DT_DEBUG_PIPE,
```
Comment (TurboGit) [inline src/common/darktable.h:278]: And here that this list must be sync with the usage display.

File: src/common/darktable.c (hunk 1 above)
Comment (jenshannoschwalm) [reply src/common/darktable.c:136]: right

File: src/common/darktable.h (hunk 2 above)
Comment (jenshannoschwalm) [reply src/common/darktable.h:278]: done

**Comment by jenshannoschwalm** [comment]
Implemented dumping pfm data for a specified module while processing the pipeline

File: src/common/darktable.c [hunk 3]
```diff
@@ -418,51 +420,111 @@ void check_resourcelevel(const char *key, int *fractions, const int level)
...
-  char fname[512];
-  snprintf(fname, sizeof (fname), "/tmp/darktable/%s/%s.pfm", modname, filename);
+  const size_t colors = (mode == DT_DUMP_PFM_MASK) ? 1 : 3;
+  const size_t p = (mode == DT_DUMP_PFM_MASK) ? 1 : 4;
+
+  char counts[32]= { 0 };
+  snprintf(counts, sizeof(counts), "%03d_", num);
+
+  char path[PATH_MAX]= { 0 };
+  snprintf(path, sizeof(path), "/tmp/darktable/%s", modname);
```
Comment (kmilos) [inline src/common/darktable.c:453]: Does GLib have some method for a temp folder location that is portable?

**Review by kmilos — CHANGES_REQUESTED** [review]

File: src/common/darktable.c (hunk 3 above)
Comment (TurboGit) [reply src/common/darktable.c:453]: And maybe even better :
https://docs.gtk.org/glib/func.mkdtemp.html

Comment (kmilos) [reply src/common/darktable.c:453]: In fact, should just copy what is already done for the log file, no?

Comment (jenshannoschwalm) [reply src/common/darktable.c:453]: Yes, my problems were
1. Do we really want a tmporary directiory? That would be good for diskspace housekeeping but would require "us to look it up"
2. Rebooting will certainly remove that ... not a problem i guess?

I propose a darktable specific tmp directory
1. it is always created when dt is started
2. it's name will be reported in the log

Comment (jenshannoschwalm) [reply src/common/darktable.c]: Done as discussed.

For my taste too many commits, if current code is ok for you i will squash.

**Comment by jenshannoschwalm** [comment]
"Squashed" into two commits with better documenting instead of simple "squash"

@TurboGit this seems ready to go now.

**Review by TurboGit — APPROVED** [review]
Works for me. Thanks!

Maybe we will need to support dumping more than a single module and so having --dump-pipe allowing a list of coma separated modules. But this can of course be done if needed later.

**Pull request merged**